src/api/eth_subscribe.rs

```diff
@@ -0,0 +1,174 @@
+//! `Eth` namespace, subscriptions
+
+use std::fmt;
+use std::marker::PhantomData;
+
+use api::Namespace;
+use futures::{Async, Future, Poll, Stream};
+use helpers::{self, CallResult};
+use rpc;
+use serde;
+use serde_json;
+use types::{BlockHeader, Filter, H256, Log};
+use {DuplexTransport, Error};
+
+/// `Eth` namespace, subscriptions
+#[derive(Debug, Clone)]
+pub struct EthSubscribe<T> {
+  transport: T,
+}
+
+impl<T: DuplexTransport> Namespace<T> for EthSubscribe<T> {
+  fn new(transport: T) -> Self
+  where
+    Self: Sized,
+  {
+    EthSubscribe { transport }
+  }
+
+  fn transport(&self) -> &T {
+    &self.transport
+  }
+}
+
+/// ID of subscription returned from `eth_subscribe`
+#[derive(Debug, Clone, Eq, Ord, PartialEq, PartialOrd)]
+pub struct SubscriptionId(String);
+
+impl From<String> for SubscriptionId {
+  fn from(s: String) -> Self {
+    SubscriptionId(s)
+  }
+}
+
+/// Stream of notifications from a subscription
+/// Given a type deserializable from rpc::Value and a subscription id, yields items of that type as
+/// notifications are delivered.
+pub struct SubscriptionStream<T: DuplexTransport, I> {
+  transport: T,
+  id: SubscriptionId,
+  rx: T::NotificationStream,
+  _marker: PhantomData<I>,
+}
+
+impl<T: DuplexTransport, I> SubscriptionStream<T, I> {
+  fn new(transport: T, id: SubscriptionId) -> Self {
+    let rx = transport.subscribe(&id);
+    SubscriptionStream {
+      transport,
+      id,
+      rx,
+      _marker: PhantomData,
+    }
+  }
+
+  /// Return the ID of this subscription
+  pub fn id(&self) -> &SubscriptionId {
+    &self.id
+  }
+
+  /// Unsubscribe from the event represented by this stream
+  pub fn unsubscribe(&self) -> CallResult<bool, T::Out> {
```

Comment:
Is it possible to consume `self` here instead of allowing to call multiple times?